DOCKER-557: fast `docker rm` without adding a transitive_state property on VM objects

misterdjules opened the pull request:
Do not wait until a VM is actually destroyed/deleted to return from a
DELETE /vms/vm_uuid request. Instead, stop the VM and at that point, do
not allow any change to this VM except for marking it as destroyed when
it's actually been destroyed.

This change also adds versioning to the current vmapi_vms bucket so that
it indexes the newly added 'zone_state' index required to be able to
search for VMs with `state === 'destroyed' && zone_state !== 'destroyed'`
when looking for "destroying" VMs.

misterdjules commented:
This change depends on https://github.com/joyent/sdc-vm-agent/pull/3.

misterdjules commented:
Here are the results of a small performance test on COAL, consisting of just measuring the time it takes to `docker rm` one `busybox` container sequentially 60 times.

Here's the distribution without this PR, with the code from master (values are in seconds):
```
➜  dragnet git:(master) ./bin/dn scan -b time_to_rm[aggr=quantize] time_to_rm_no_fix                                          

           value  ------------- Distribution ------------- count
               0 |                                         0
               1 |                                         0
               2 |                                         0
               4 |@@@                                      5
               8 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@    55
              16 |                                         0
➜  dragnet git:(master)
```
And here's the distribution with the changes from this PR:
```
➜  dragnet git:(master) ./bin/dn scan -b time_to_rm[aggr=quantize] time_to_rm_with_fix

           value  ------------- Distribution ------------- count
               0 |                                         0
               1 |@@                                       4
               2 |                                         0
               4 |@@@@@@@@@@@@@@@                          25
               8 |@@@@@@@@@@@@@@@@@@@@@@                   36
              16 |                                         0
➜  dragnet git:(master)
```

As we can see, with the code from this PR there's a lot more invocations of `docker rm` that fall into the 4 seconds bucket.

What's interesting is that the values are actually always slightly higher than 5 seconds or slightly higher than 10 seconds. The reason is that the code that polls for job results runs at 5 seconds intervals. So what happens with the changes in this PR is that we're slightly faster with small containers and we get lucky more often.

This polling behavior is in itself something that should be improved, but it's not in the scope of this PR.

There's also still more tests to run with containers for which the destroy operation (not stop) takes longer. I'll post results of such tests soon.

misterdjules commented:
@joshwilsdon Do you have some time to review this PR?

misterdjules commented:
@joshwilsdon I've been able to confirm that deleting VMs with `archive_on_delete: true` with large core files in `/zones/zone-id/cores` is faster with the changes in this PR, because it doesn't wait for the VM to be destroyed (and thus the core files archived) to return the response. The deletion time is constant, as one would expect, whereas without these changes it is longer the more files have to be archived and the bigger they are.

misterdjules commented:
@joshwilsdon Thank you for the first code review, I updated this PR with a second commit that addresses your comments. I will squash all commits into one when we're ready to merge it.

joshwilsdon commented on test/vms.destroy.test.js:
```diff
@@ -0,0 +1,532 @@
+    });
+};
+
+exports.post_on_destroying_vm_must_fail = function (t) {
```
We're checking all these POST actions, should we also not validate that the behavior is correct in the case that someone tries a DELETE on a VM that's 'destroying'?

misterdjules replied:
There's already the [`destroy_destructible_vm`](https://github.com/joyent/sdc-vmapi/pull/11/files#diff-bf4e451cca98fd1a17105a1393a19e4cR408) test that deletes the test VM when it's in the `destroying` state and its `indestructible_zoneroot` property has been set to `false`.

Did you have another test in mind?

joshwilsdon commented on test/vms.destroy.test.js:
```diff
@@ -0,0 +1,532 @@
+};
+
+exports.remove_leftover_test_vms = function (t) {
+    async.each(leftoverTestVms, function (testVm, nextVm) {
```
Is it better to use vasync instead of async here? I think in general we're trying to move away from async and to vasync... But if it's hard to do in this module it's fine to do that separately later.

misterdjules replied:
Excellent point, will fix!

joshwilsdon commented on package.json:
```diff
@@ -5,24 +5,25 @@
   "private": true,
   "dependencies": {
-    "libuuid": "0.1.2",
+    "amqp": "git://github.com/postwait/node-amqp.git#2a5b1e9d2cd5838991ea378251f186c13ba79f97",
```
Thanks for sorting these. :)

misterdjules replied:
`npm` has been reordering dependencies on `--save` for a while now, so I haven't done anything myself to sort this list alphabetically. If that's not something we want, please let me know.

misterdjules commented:
@joshwilsdon Thank you for the second review :) I believe I have addressed most of your suggestions in the most recent commit of this PR.

misterdjules commented:
@joshwilsdon Just for the record, I've ran several migrations successfully with these changes as well as https://github.com/joyent/sdc-vm-agent/pull/3 after setting up another CN on my COAL.

It's obviously not a sign that the race conditions we talked about are fixed by this PR, but it's an indication that this PR at least doesn't seem to break the migration use case in any obvious way.

misterdjules commented:
@joshwilsdon Pushed [another commit](https://github.com/misterdjules/sdc-vmapi/commit/43d766624a49aa9d47d951c23c5f69c593cb5536) that I believe addresses all your comments, thank you again!

joshwilsdon commented on lib/apis/moray.js:
```diff
@@ -493,6 +488,34 @@ Moray.prototype._createSearch = function (params, cb) {
+ * @param  {string}     uuid uuid of the server whos vms should be fetched.
+ * @param  {Function}   cb  function of the form function(err, vms) {...}
+ */
+Moray.prototype.listVmsForServer = function listVmsForServer(uuid, cb) {
```
The biggest remaining question I have is whether we can change Moray.prototype.listVmsForServer such that we don't grab VMs which have zone_state=destroyed && state=destroyed. The reason for this is that as far as I understand it, the current code will load all VMs from the beginning of history. Which, especially for CNs handling docker VMs, could be a very large amount. We'll likely hit limits on that and need to do paging and this will just get slower and slower over time. If we ignore the destroyed && destroyed, we would only be looking at the current set of active VMs and those being destroyed. 

It seems like this might be possible since validStateChange() only checks for zone_state !== 'destroyed' which seems like it will be fine?

misterdjules replied:
@joshwilsdon The performance issue is a very good point. As per our discussions, we do not load all VMs from moray. Instead, we load all VMs that are passed in the payload of the PUT /vms request. It's still a lot, and the performance is not great but:

1. It's bounded by the number of "live" VMs that can fit on a CN.
2. By loading each VM passed as the request payload individually and [in parallel](https://github.com/joyent/sdc-vmapi/pull/11/files#diff-8491b9ad8bfb1f56e7e6c552ad1959e6R1219), and by updating them in parallel too, the performance according to my tests is actually slightly better than the previous implementation.

We also discussed a new design that would make the PUT /vms endpoint obsolete, and would get rid of the performance issues by design. I believe the latest update of this PR gives us an acceptable intermediate step towards that goal.

misterdjules commented:
@joshwilsdon Thanks! Addressed all your comments and added a test for the `'manual_override'` state in `test/vms.state_manual_override.test.js`.

misterdjules commented:
Closing in favor of https://github.com/joyent/sdc-vmapi/pull/16.
